algorithms/unit_tests/TestSort.hpp

```diff
@@ -402,6 +403,339 @@ void test_sort_integer_overflow() {
+    for (unsigned j = 0; j < keysAndValues.size(); j++) {
+      keysAndValues[j].first  = keysHost(offsetsHost(i) + j);
+      keysAndValues[j].second = valuesHost(offsetsHost(i) + j);
+    }
+    if (customCompare) {
+      std::sort(keysAndValues.begin(), keysAndValues.end(),
+                [](const KV& a, const KV& b) { return a.first > b.first; });
+    } else {
+      std::sort(keysAndValues.begin(), keysAndValues.end(),
+                [](const KV& a, const KV& b) { return a.first < b.first; });
+    }
+    // Copy back from pairs to views
+    for (unsigned j = 0; j < keysAndValues.size(); j++) {
+      keysHost(offsetsHost(i) + j)   = keysAndValues[j].first;
+      valuesHost(offsetsHost(i) + j) = keysAndValues[j].second;
+    }
+  }
+  if (useTeams) {
+    int vectorLen = std::min<int>(4, TeamPol::vector_length_max());
+    TeamPol policy(narray, Kokkos::AUTO(), vectorLen);
+    Kokkos::parallel_for(
+        policy, TeamSortByKeyFunctor<ExecutionSpace, KeyViewType, ValueViewType,
+                                     OffsetViewType>(keys, values, offsets,
+                                                     customCompare));
+  } else {
+    ThreadSortByKeyFunctor<ExecutionSpace, KeyViewType, ValueViewType,
+                           OffsetViewType>
+        functor(keys, values, offsets, customCompare);
+    int vectorLen = std::min<int>(4, TeamPol::vector_length_max());
+    TeamPol dummy(1, Kokkos::AUTO(), vectorLen);
+    int teamSize =
+        dummy.team_size_recommended(functor, Kokkos::ParallelForTag());
+    int numTeams = (narray + teamSize - 1) / teamSize;
+    Kokkos::parallel_for(TeamPol(numTeams, teamSize, vectorLen), functor);
+  }
+  auto keysOut = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), keys);
+  auto valuesOut =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), values);
+  // First, compare keys since they will always match exactly
+  for (unsigned i = 0; i < keys.extent(0); i++) {
+    EXPECT_EQ(keysOut(i), keysHost(i));
+  }
+  // Kokkos::sort_by_key_X is not stable, so if a key happens to
+  // appear more than once, the order of the values may not match exactly.
+  // But the set of values for a given key should be identical.
+  unsigned keyStart = 0;
+  while (keyStart < keys.extent(0)) {
+    KeyType key     = keysHost(keyStart);
+    unsigned keyEnd = keyStart + 1;
+    while (keyEnd < keys.extent(0) && keysHost(keyEnd) == key) keyEnd++;
+    std::unordered_multiset<ValueType> correctVals;
+    std::unordered_multiset<ValueType> outputVals;
+    for (unsigned i = keyStart; i < keyEnd; i++) {
+      correctVals.insert(valuesHost(i));
+      outputVals.insert(valuesOut(i));
+    }
+    // Check one value at a time that they match
+    for (auto it = correctVals.begin(); it != correctVals.end(); it++) {
+      ValueType val = *it;
+      EXPECT_TRUE(outputVals.find(val) != outputVals.end());
+      EXPECT_EQ(correctVals.count(val), outputVals.count(val));
```

Comment:
Try to produce more meaningful error messages when these fail.
You may use the `operator<<` on the assertion or refactor to have individual unit tests.